client-rpc/src/rpc/websocket_rpc.rs

```diff
@@ -0,0 +1,154 @@
+use client_common::{PrivateKey, PublicKey, Storage};
+use client_core::WalletClient;
+use client_index::BlockHandler;
+use futures::future::Future;
+use futures::sink::Sink;
+use futures::stream::Stream;
+use futures::sync::mpsc;
+use std::thread;
+use websocket::result::WebSocketError;
+use websocket::{ClientBuilder, OwnedMessage};
+
+// this handles low level network connection
+// packet processing and core works are done in websocket_core
+// it uses channel to communicate with core
+// through send queue
+pub const CMD_SUBSCRIBE: &str = r#"
+    {
+        "jsonrpc": "2.0",
+        "method": "subscribe",
+        "id": "subscribe_reply",
+        "params": {
+            "query": "tm.event='NewBlock'"
+        } 
+    }"#;
+pub const CMD_BLOCK: &str = r#"
+    {
+        "method": "block",
+        "jsonrpc": "2.0",
+        "params": [ "2" ],
+        "id": "block_reply"
+    }"#;
+pub const CMD_STATUS: &str = r#"
+    {
+        "method": "status",
+        "jsonrpc": "2.0",
+        "params": [ ],
+        "id": "status_reply"
+    }"#;
+
+type MyQueue = std::sync::mpsc::Sender<OwnedMessage>;
+
+#[derive(Clone, Debug)]
+pub struct WalletInfo {
+    pub name: String,
+    pub staking_addresses: Vec<StakedStateAddress>,
+    pub view_key: PublicKey,
+    pub private_key: PrivateKey,
+}
+pub type WalletInfos = Vec<WalletInfo>;
+
+// constanct connection
+// using ws://localhost:26657/websocket
+pub struct WebsocketRpc {
+    pub core: Option<MyQueue>,
+    websocket_url: String,
+    my_sender: Option<mpsc::Sender<OwnedMessage>>,
+    my_receiver: Option<mpsc::Receiver<OwnedMessage>>,
+}
+
+impl WebsocketRpc {
+    pub fn new(websocket_url: String) -> Self {
+        Self {
+            core: None,
+            websocket_url,
+            my_sender: None,
+            my_receiver: None,
+        }
+    }
+
+    // launch core thread
+    pub fn run<
+        S: Storage + 'static,
+        C: Client + 'static,
+        H: BlockHandler + 'static,
+        T: WalletClient + 'static,
+    >(
+        &mut self,
+        wallets: WalletInfos,
+        client: C,
+        storage: S,
+        block_handler: H,
+        wallet_client: T,
+    ) {
+        println!("Connecting to {}", self.websocket_url);
+        let channel = mpsc::channel(0);
+        // tx, rx
+        let (channel_tx, channel_rx) = channel;
+        self.my_sender = Some(channel_tx.clone());
+        self.my_receiver = Some(channel_rx);
+
+        let mut core = WebsocketCore::new(
+            channel_tx.clone(),
+            storage,
+            client,
+            block_handler,
+            wallets,
+            wallet_client,
+        );
+        // save send_queue to communicate with core
+        self.core = Some(core.get_queue());
+
+        let _child = thread::spawn(move || {
+            core.start();
+        });
+
+        assert!(self.core.is_some());
+    }
+
+    // activate tokio websocket 
+    pub fn run_network(&mut self) {
```

Comment:
I think the whole code in this function is boilerplate and can be removed if we use [jsonrpc-ws-client](https://crates.io/crates/jsonrpc-ws-client).

I think it is a good idea to use crates which are well tested and used by many. Maybe we can do this in a future PR. Any comments @tomtau?

Reply:
client-rpc already uses jsonrpc-core-client (or does it? it has it in cargo dependencies at least) which abstracts over the transport (tls, http or ws) -- so it'll make sense to use jsonrpc-ws-client.

Reply:
but yeah, using jsonrpc-ws-client should reduce some of the concurrency boilerplate

Reply:
core and rpc networking communicates with channel,
that means calling function is not allowed, only serialized strings.

Reply:
https://crates.io/crates/jsonrpc-ws-client handles that for you

Reply:
i'm making test code for this.

Reply:
request makes channel for each request, and
subscription handling is not documented well.
so after researching issues, i'll comment the more detail.